index.js

```diff
@@ -173,22 +188,28 @@ DB.prototype.del = function (key, opts, cb) {
   })
 }
 
-DB.prototype._del = function (key, opts, cb) {
+// OsmVersion, Opts -> [OsmBatchOp]
+DB.prototype._getDocumentDeletionBatchOps = function (key, opts, cb) {
   var self = this
-  self.kv.get(key, function (err, values) {
+  self.kv.get(key, function (err, docs) {
     if (err) return cb(err)
-    // Filter deletions & map to old expected format.
-    values = filterObj(values, function (key, value) {
-      return !value.deleted
-    })
-    values = mapObj(values, function (key, value) {
-      return value.value
+
+    docs = mapObj(docs, function (version, value) {
+      if (value.deleted) {
+        return {
+          id: key,
+          version: version,
+          deleted: true
+        }
+      } else {
+        return value.value
+      }
     })
 
     var fields = {}
-    var links = opts.keys || Object.keys(values)
+    var links = opts.keys || Object.keys(docs)
     links.forEach(function (ln) {
-      var v = values[ln] || {}
+      var v = docs[ln] || {}
```

Comment:
We need to handle when `opts.links` are not the current head(s), e.g. when `Object.keys(docs) !== opts.links`. This can happen when db state changes between when the client read and when the write the delete. We need an additional lookup here.